[AArch64] Add SME peephole optimizer pass

sdesmalen-arm opened the pull request:
This pass removes back-to-back smstart/smstop instructions
to reduce the number of streaming mode changes in a function.

The implementation as proposed doesn't aim to solve all problems
yet and suggests a number of cases that can be optimized in the
future.

sdesmalen-arm committed:
[AArch64] Add SME peephole optimizer pass

This pass removes back-to-back smstart/smstop instructions
to reduce the number of streaming mode changes in a function.

The implementation as proposed doesn't aim to solve all problems
yet and suggests a number of cases that can be optimized in the
future.

sdesmalen-arm committed:
Precommit test

jroelofs reviewed (approved):
nice!

davemgreen commented on llvm/lib/Target/AArch64/SMEPeepholeOpt.cpp:
```diff
@@ -0,0 +1,216 @@
+        Prev = nullptr;
+        Changed = true;
+      } else {
+        Reset();
```
I might not be following this correctly, but if each block only has one smstart/smstop, does this ever get reached, and so does HasRemainingSMChange even get set? Is it relying on always having another instruction in the block (say a terminator)?

Like I said I may not understand the reasoning, and the pass otherwise looks good.

sdesmalen-arm replied:
In `isMatchingStartStopPair` we test if these instructions are a matching `smstart, smstop` or `smstop, smstart` pair, based on the operands to this pseudo. (we don't want the algorithm to fold away `smstart sm, smstart za` for example). If we don't find a matching pseudo, it ends up in this block and the algorithm resets.

davemgreen replied:
I see. My point was maybe quite contrived. If have a block like:
```
bb:
  smstart

```
It doesn't have anything else in it (maybe a copy or another "allowed" instruction), and no terminator. There would be another block with the corresponding smstop, and another block with a smstart/smtop pair that gets removed (to set `Changed=true`). In the block above it would start out as `HasRemainingSMChange = false;`, look at the smstart to set `if (!Prev) Prev = &MI;`, then return because there were no changes. Nothing set `HasRemainingSMChange`, so the runOnMachineFunction function sets `AFI->setHasStreamingModeChanges(FunctionHasRemainingSMChange);` (= false). It has to be a bit contrived to set `Changed` but not `FunctionHasRemainingSMChange`.

sdesmalen-arm replied:
Thanks for clarifying. Yes, that's an unusual case that wouldn't ever happen with the way the smstart/smstop nodes are emitted, but I agree that the pass should be written with this case in mind. I've fixed the issue but I am unable to write an explicit test for it.

davemgreen commented on llvm/lib/Target/AArch64/SMEPeepholeOpt.cpp:
```diff
@@ -0,0 +1,216 @@
+      break;
+    case AArch64::COPY: {
+      // Permit copies of 32 and 64-bit registers.
+      if (!MI.getOperand(1).isReg()) {
```
Should this check the first operand too, just in case in is a cross-register-bank gpr->fpr copy?

sdesmalen-arm replied:
I found that we were a bit too restrictive on the COPY nodes, so I've relaxed the restrictions a bit more, e.g. FPR copies should be fine, although I've stayed conservative on PPR/ZPR copies, as I can't prove they're always safe.

I also did some minor refactoring of the switch statement so that we don't have to analyse COPY instructions (or any other instructions) when we're not trying to match a sequence.

sdesmalen-arm committed:
Relax restrictions for COPY and refactor

The restrictions for the COPY nodes were too strict. I couldn't
come up with any tests or cases where a COPY on its own would
result in any issues.

Also refactored the code a bit so that we don't need to do any
analysis on COPY nodes when the algorithm isn't trying to match a
'smstart/smstop' sequence.

aemerson reviewed (approved):
LGTM, thanks for this.

sdesmalen-arm committed:
Fix issue where HasStreamingModeChanges may be set to 'false' incorrectly

davemgreen reviewed (approved):
Thanks. This looks like a useful pass.

sdesmalen-arm committed:
Fix failing test before committing
